<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>Teachingmaterial by lgatto</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/github-light.css">
    <meta name="viewport" content="width=device-width">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1>Teachingmaterial</h1>
        <p>Various teaching material</p>

        <p class="view"><a href="https://github.com/lgatto/TeachingMaterial">View the Project on GitHub <small>lgatto/TeachingMaterial</small></a></p>


        <ul>
          <li><a href="https://github.com/lgatto/TeachingMaterial/zipball/master">Download <strong>ZIP File</strong></a></li>
          <li><a href="https://github.com/lgatto/TeachingMaterial/tarball/master">Download <strong>TAR Ball</strong></a></li>
          <li><a href="https://github.com/lgatto/TeachingMaterial">View On <strong>GitHub</strong></a></li>
        </ul>
      </header>
      <section>
        <h1>
<a id="teachingmaterial" class="anchor" href="#teachingmaterial" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>TeachingMaterial</h1>

<p>This repository is an aggregator for various
<a href="http://www.r-project.org/">R</a>, <code>make</code> and <code>git</code>/<code>github</code> teaching
material.  Most of the courses are taught at the University of
Cambridge, UK, and some have been adapted and exported outside. We
would also like to acknowledge contributions from
<a href="http://www.software.ac.uk/about/people/aleksandra-pawlik">Aleksandra Pawlik</a>,
Software Sustainability Institute,
<a href="http://www.rglab.org/">Raphael Gottardo</a>, Fred Hutchinson Cancer
Research Center and <a href="http://biostat.wisc.edu/%7Ekbroman/">Karl Broman</a>,
University of Wisconsin-Madison.</p>

<p>Each material subdirectory has its own repository; <code>TeachingMaterial</code>
aggregates a snapshot as a central entry point.  Aggregation is done
using <code>git-subtree</code> (see the
<a href="https://github.com/lgatto/TeachingMaterial/wiki/TM-Administration">administration page</a>
for details).  The local copies linking to external repositories are
prefixed with an underscore.</p>

<p>Unless otherwise stated, all material is licensed under a
<a href="http://creativecommons.org/licenses/by-sa/3.0/">Creative Commons Attribution-ShareAlike 3.0 License</a>.
This means you are free to copy, distribute and transmit the work,
adapt it to your needs as long as you cite its origin and, if you do
redistribute it, do so under the same license.</p>

<p>See also the
<a href="https://github.com/lgatto/teachingmaterial/wiki"><code>TeachingMaterial</code> wiki</a>
for meta-information about the repository and general <code>R</code> installation
material and links.</p>

<p>If you like this material and/or this initiative, do not hesitate to
let us know by starring the repo, tweeting about it and sharing it
with your colleagues.</p>

<h2>
<a id="material" class="anchor" href="#material" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Material</h2>

<h3>
<a id="r-debugging-and-robust-programming" class="anchor" href="#r-debugging-and-robust-programming" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>R debugging and robust programming</h3>

<ul>
<li>Description: A 2-day workshop taught on the 25-26 February 2016 at
the EMBL, Heidelberg. The course aims at teaching participants
debugging techniques and good practice in writing reliable, robust
code.</li>
<li>Author: <a href="https://github.com/lgatto">Laurent Gatto</a>, based on
previous content by Laurent Gatto and Robert Stojnic, and
<a href="http://adv-r.had.co.nz/"><em>Advanced R</em></a>, by Hadley Wickham.</li>
<li>Original repository: <a href="https://github.com/lgatto/2016-02-25-adv-programming-EMBL">https://github.com/lgatto/2016-02-25-adv-programming-EMBL</a>
</li>
<li>Content: Part I: Coding style(s), Interactive use and programming,
Environments, Tidy data, Computing on the language. Part II:
Functions, Robust programming with functions, Scoping, Closures,
High-level functions, Vectorisation. Part III: Defensive
programming, Debbugging: techniques and tools, Condition handling:
try/tryCatch, Unit testing. Part IV: Benchmarking, Profiling,
Optimisation, Memory, Rcpp.</li>
<li>More details: <a href="https://github.com/lgatto/2016-02-25-adv-programming-EMBL/blob/master/README.md">https://github.com/lgatto/2016-02-25-adv-programming-EMBL/blob/master/README.md</a>
</li>
</ul>

<h3>
<a id="rbc" class="anchor" href="#rbc" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>rbc</h3>

<ul>
<li>Description:
<a href="http://github.com/lgatto/rbc">Software carpentry R bootcamp</a>, Jan
7-8, 2014, Cambridge, UK and 6-7 Nov 2014, Zurich, Switzerland.</li>
<li>Authors: <a href="http://www.damtp.cam.ac.uk/user/sje30/">Stephen Eglen</a>,
<a href="https://github.com/lgatto">Laurent Gatto</a>, Robert Stojnić and
<a href="http://www.software.ac.uk/about/people/aleksandra-pawlik">Aleksandra Pawlik</a>
</li>
<li>Original repository: <a href="https://github.com/lgatto/rbc/">https://github.com/lgatto/rbc/</a>
</li>
<li>Content: <code>R</code> programming, plotting, <code>git</code>/<code>github</code> (via
<a href="http://software-carpentry.org/">software carpentry</a>), <code>make</code>,
<code>shell</code> and <code>knitr</code>, profiling, testing, debugging.</li>
</ul>

<h3>
<a id="spr" class="anchor" href="#spr" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>spr</h3>

<ul>
<li>Description: Scientific Programming with <code>R</code>, MPhil in Computational Biology</li>
<li>Author: <a href="http://www.damtp.cam.ac.uk/user/sje30/">Stephen Eglen</a> with contributions from <a href="https://github.com/lgatto">Laurent Gatto</a>
</li>
<li>Original repository: <a href="https://github.com/lgatto/spr">https://github.com/lgatto/spr</a>
</li>
<li>More details: <a href="https://github.com/lgatto/spr#readme">https://github.com/lgatto/spr#readme</a>
</li>
</ul>

<h3>
<a id="biostat-578" class="anchor" href="#biostat-578" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Biostat-578</h3>

<ul>
<li>Description: Bioinformatics for Big Omics Data</li>
<li>Author: <a href="http://www.rglab.org/">Raphael Gottardo</a>, Fred Hutchinson Cancer Research Center</li>
<li>Original repository: <a href="https://github.com/raphg/Biostat-578">https://github.com/raphg/Biostat-578</a>
</li>
<li>More details: <a href="https://github.com/raphg/Biostat-578/blob/master/README.md">https://github.com/raphg/Biostat-578/blob/master/README.md</a>
</li>
</ul>

<h3>
<a id="rbioc-proteomics" class="anchor" href="#rbioc-proteomics" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>rbioc-proteomics</h3>

<ul>
<li>Using R/Bioconductor for MS-based proteomics data analysis</li>
<li>Author: <a href="https://github.com/lgatto">Laurent Gatto</a>
</li>
<li>Original repository: <a href="https://github.com/lgatto/rbioc-proteomics">https://github.com/lgatto/rbioc-proteomics</a>
</li>
<li>More details: <a href="https://github.com/lgatto/rbioc-proteomics#readme">https://github.com/lgatto/rbioc-proteomics#readme</a>
</li>
</ul>

<h3>
<a id="github_tutorial" class="anchor" href="#github_tutorial" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>github_tutorial</h3>

<ul>
<li>Description: <code>git</code>/<code>github</code> guide</li>
<li>Original repository: <a href="https://github.com/kbroman/github_tutorial">https://github.com/kbroman/github_tutorial</a>
</li>
<li>Author: <a href="http://biostat.wisc.edu/%7Ekbroman/">Karl Broman</a>, University of Wisconsin-Madison</li>
<li>View it here: <a href="http://kbroman.github.io/github_tutorial/">http://kbroman.github.io/github_tutorial/</a>
</li>
</ul>

<h3>
<a id="minimal_make" class="anchor" href="#minimal_make" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>minimal_make</h3>

<ul>
<li>Description: minimal tutorial on <code>make</code>
</li>
<li>Original repository: <a href="https://github.com/kbroman/minimal_make">https://github.com/kbroman/minimal_make</a>
</li>
<li>Author: <a href="http://biostat.wisc.edu/%7Ekbroman/">Karl Broman</a>, University of Wisconsin-Madison</li>
<li>View it here: <a href="http://kbroman.github.io/minimal_make/">http://kbroman.github.io/minimal_make/</a>
</li>
</ul>

<h3>
<a id="quickpackage" class="anchor" href="#quickpackage" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>QuickPackage</h3>

<ul>
<li>Description: Two brief overviews of <code>R</code> package creation</li>
<li>Author: <a href="https://github.com/lgatto">Laurent Gatto</a>
</li>
<li>Original repository: <a href="https://github.com/lgatto/QuickPackage">https://github.com/lgatto/QuickPackage</a>
</li>
<li>More details: <a href="https://github.com/lgatto/QuickPackage#readme">https://github.com/lgatto/QuickPackage#readme</a>
</li>
<li>Download the <a href="https://github.com/lgatto/QuickPackage/blob/master/QuickPackage.pdf?raw=true">QuickPackage</a> and <a href="https://github.com/lgatto/QuickPackage/blob/master/QuickPackageAndMore.pdf?raw=true">QuickPackageAndMore</a> slides (pdf)</li>
</ul>

<h3>
<a id="r-package-development" class="anchor" href="#r-package-development" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>R package development</h3>

<ul>
<li>Description: Developing, documenting and testing an <code>R</code> package</li>
<li>Author: <a href="https://github.com/lgatto">Laurent Gatto</a> and Robert Stojnić</li>
<li>Original repository: <a href="https://github.com/lgatto/RPackageDevelopment">https://github.com/lgatto/RPackageDevelopment</a>
</li>
<li>More details: <a href="https://github.com/lgatto/RPackageDevelopment#readme">https://github.com/lgatto/RPackageDevelopment#readme</a>
</li>
<li>Download the <a href="https://github.com/lgatto/RPackageDevelopment/blob/master/rpd.pdf?raw=true">pdf</a>
</li>
</ul>

<h3>
<a id="benchmarking-profiling-and-optimisation" class="anchor" href="#benchmarking-profiling-and-optimisation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Benchmarking, profiling and optimisation</h3>

<ul>
<li>Description: Benchmarking, profiling and optimisation</li>
<li>Author: <a href="https://github.com/lgatto">Laurent Gatto</a>
</li>
<li>Original repository: <a href="https://github.com/lgatto/R-bmark-prof-optim">https://github.com/lgatto/R-bmark-prof-optim</a>
</li>
<li>More details: <a href="https://github.com/lgatto/R-bmark-prof-optim#readme">https://github.com/lgatto/R-bmark-prof-optim#readme</a>
</li>
<li>Read the <a href="https://github.com/lgatto/R-bmark-prof-optim/blob/master/bmark-prof-optim.md">material</a>
</li>
</ul>

<h3>
<a id="rbasics" class="anchor" href="#rbasics" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>RBasics</h3>

<ul>
<li>Description: A introduction to <code>R</code> for knowledgeable bioinformaticians. Used as <code>R</code> intro lecture for the <a href="http://marray.economia.unimi.it/">CSAMA</a> workshop.</li>
<li>Author: <a href="https://github.com/lgatto">Laurent Gatto</a>
</li>
<li>Original repository: <a href="https://github.com/lgatto/RBasics">https://github.com/lgatto/RBasics</a>
</li>
<li>More details: <a href="https://github.com/lgatto/RBasics#readme">https://github.com/lgatto/RBasics#readme</a>
</li>
<li>Download the <a href="https://github.com/lgatto/RBasics/blob/master/R-Basics.pdf?raw=true">pdf</a>
</li>
</ul>

<h3>
<a id="rintro" class="anchor" href="#rintro" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>RIntro</h3>

<ul>
<li>Description: A 1 day introductionary course, developed specifically for Microarray Analysis using <code>R</code> and Bioconductor.</li>
<li>Author: <a href="https://github.com/lgatto">Laurent Gatto</a>
</li>
<li>Original repository: <a href="https://github.com/lgatto/RIntro">https://github.com/lgatto/RIntro</a>
</li>
<li>More details: <a href="https://github.com/lgatto/RIntro#readme">https://github.com/lgatto/RIntro#readme</a>
</li>
<li>Download the <a href="https://github.com/lgatto/RIntro/blob/master/RIntro.pdf?raw=true">pdf</a>
</li>
</ul>

<h3>
<a id="basicr" class="anchor" href="#basicr" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>basicr</h3>

<ul>
<li>Description: Material for the 2-day <code>R</code> introduction taught at Cambridge's <a href="http://www.training.cam.ac.uk/gsls/course/gsls-rintro">Graduate School of Life Sciences Training</a>.</li>
<li>Authors: Robert Stojnic, <a href="https://github.com/lgatto">Laurent Gatto</a>, Rob Foy, David Molnar and John Davey, based on original slides by Ian Roberts and Robert Stojnic.</li>
<li>Original repository: <a href="https://github.com/johnomics/basicr/">https://github.com/johnomics/basicr/</a>
</li>
<li>Course page: <a href="http://logic.sysbiol.cam.ac.uk/teaching/Rcourse/">http://logic.sysbiol.cam.ac.uk/teaching/Rcourse/</a>
</li>
<li>Download the slides for <a href="https://github.com/johnomics/basicr/raw/master/Basic_R_Day_1_slides.pdf">day 1</a> and <a href="https://github.com/johnomics/basicr/raw/master/Basic_R_Day_2_slides.pdf">day 2</a> in pdf.</li>
</ul>

<h3>
<a id="r-functional-programming" class="anchor" href="#r-functional-programming" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>R functional programming</h3>

<ul>
<li>Description: A short topic section on functional programming in <code>R</code>
</li>
<li>Author: <a href="https://github.com/lgatto">Laurent Gatto</a>
</li>
<li>Original repository: <a href="https://github.com/lgatto/R-functional-programming">https://github.com/lgatto/R-functional-programming</a>
</li>
<li>More details: <a href="https://github.com/lgatto/R-functional-programming#readme">https://github.com/lgatto/R-functional-programming#readme</a>
</li>
<li>Download the <a href="https://github.com/lgatto/R-functional-programming/blob/master/functional-programming.pdf?raw=true">pdf</a>
</li>
</ul>

<h3>
<a id="r-vectorisation" class="anchor" href="#r-vectorisation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>R vectorisation</h3>

<ul>
<li>Description: A short topic section on vectorisation in <code>R</code>
</li>
<li>Author: <a href="https://github.com/lgatto">Laurent Gatto</a>
</li>
<li>Original repository: <a href="https://github.com/lgatto/R-vectorisation">https://github.com/lgatto/R-vectorisation</a>
</li>
<li>More details: <a href="https://github.com/lgatto/R-vectorisation#readme">https://github.com/lgatto/R-vectorisation#readme</a>
</li>
<li>Download the <a href="https://github.com/lgatto/R-vectorisation/blob/master/vectorisation.pdf?raw=true">pdf</a>
</li>
</ul>

<h3>
<a id="r-debugging" class="anchor" href="#r-debugging" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>R debugging</h3>

<ul>
<li>Description: A short topic section on debugging in <code>R</code>
</li>
<li>Author: <a href="https://github.com/lgatto">Laurent Gatto</a> and Robert Stojnić</li>
<li>Original repository: <a href="https://github.com/lgatto/R-debugging">https://github.com/lgatto/R-debugging</a>
</li>
<li>More details: <a href="https://github.com/lgatto/R-debugging#readme">https://github.com/lgatto/R-debugging#readme</a>
</li>
<li>Download the <a href="https://github.com/lgatto/R-debugging/blob/master/debugging.pdf?raw=true">pdf</a>
</li>
</ul>

<h3>
<a id="r-parallel" class="anchor" href="#r-parallel" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>R parallel</h3>

<ul>
<li>Description: A short topic section on parallal computing in <code>R</code>
</li>
<li>Author: <a href="https://github.com/lgatto">Laurent Gatto</a> and Robert Stojnić</li>
<li>Original repository: <a href="https://github.com/lgatto/R-parallel">https://github.com/lgatto/R-parallel</a>
</li>
<li>More details: <a href="https://github.com/lgatto/R-parallel#readme">https://github.com/lgatto/R-parallel#readme</a>
</li>
<li>Download the <a href="https://github.com/lgatto/R-parallel/blob/master/parallel.pdf?raw=true">pdf</a>
</li>
</ul>

<h3>
<a id="r-object-oriented-programming" class="anchor" href="#r-object-oriented-programming" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>R object oriented programming</h3>

<ul>
<li>Description: Covers S3, S4 and S4 Reference Classes OO programming using DNA/RNA sequence data manipulation as a working example.</li>
<li>Author: <a href="https://github.com/lgatto">Laurent Gatto</a> and Robert Stojnić</li>
<li>Original repository: <a href="https://github.com/lgatto/roo">https://github.com/lgatto/roo</a>
</li>
<li>More details: <a href="https://github.com/lgatto/roo#readme">https://github.com/lgatto/roo#readme</a>
</li>
<li>Download the <a href="https://github.com/lgatto/roo/blob/master/roo.pdf?raw=true">pdf</a>
</li>
</ul>

<h3>
<a id="short-s4-tutorial" class="anchor" href="#short-s4-tutorial" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Short S4 tutorial</h3>

<ul>
<li>Description: A practical S4 object-oriented programming tutorial using microarrays as a working example.</li>
<li>Author: <a href="https://github.com/lgatto">Laurent Gatto</a> </li>
<li>Original repository: <a href="https://github.com/lgatto/S4-tutorial">https://github.com/lgatto/S4-tutorial</a>
</li>
<li>More details: <a href="https://github.com/lgatto/S4-tutorial#readme">https://github.com/lgatto/S4-tutorial#readme</a>
</li>
<li>Download the <a href="https://github.com/lgatto/S4-tutorial/blob/master/S4-tutorial.pdf?raw=true">pdf</a>
</li>
</ul>

<h3>
<a id="r-programming-tutorial" class="anchor" href="#r-programming-tutorial" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>R programming tutorial</h3>

<ul>
<li>Description: A tutorial on R programming of intermediate level, focusing on some aspects of functional programming, profiling, testing, debugging and parallelisation. Used as more advanced <code>R</code> programming lecture during the <a href="http://marray.economia.unimi.it/">CSAMA</a> workshop.</li>
<li>Author: <a href="https://github.com/lgatto">Laurent Gatto</a> </li>
<li>Original repository: <a href="https://github.com/lgatto/R-programming">https://github.com/lgatto/R-programming</a>
</li>
<li>More details: <a href="https://github.com/lgatto/R-programming#readme">https://github.com/lgatto/R-programming#readme</a>
</li>
<li>Download the <a href="https://github.com/lgatto/R-programming/blob/master/R-programming.pdf?raw=true">pdf</a>
</li>
</ul>

<h3>
<a id="r-and-cc" class="anchor" href="#r-and-cc" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>R and C/C++</h3>

<ul>
<li>Description: Writing and calling <code>C</code>/<code>C++</code> code for/from <code>R</code>
</li>
<li>Author:  <a href="https://github.com/lgatto">Laurent Gatto</a> </li>
<li>Many examples from the <a href="http://gallery.rcpp.org/">Rcpp gallery</a> and 
from <a href="http://dirk.eddelbuettel.com/bio/presentations.html">Dirk's slides</a>.</li>
<li>Original repository: <a href="https://github.com/lgatto/rccpp">https://github.com/lgatto/rccpp</a>
</li>
<li>More details: <a href="https://github.com/lgatto/rccpp#readme">https://github.com/lgatto/rccpp#readme</a>
</li>
<li>Download the <a href="https://github.com/lgatto/rccpp/blob/master/rccpp.pdf?raw=true">pdf</a>
</li>
</ul>

<h3>
<a id="visualisation" class="anchor" href="#visualisation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>visualisation</h3>

<ul>
<li>Description: visualising data using <code>ggplot2</code> (more to come soon)</li>
<li>Author: Mark Dunning</li>
<li>Original repository: <a href="https://github.com/lgatto/visualisation">https://github.com/lgatto/visualisation</a>
</li>
<li>Download the <a href="https://github.com/lgatto/visualisation/blob/master/ggplot2_cambr28oct2013.pdf?raw=true">pdf</a>
</li>
</ul>

<h3>
<a id="sequences" class="anchor" href="#sequences" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>sequences</h3>

<ul>
<li>Description: Educational package used in <code>R</code> to illustrate OO programming and package development</li>
<li>Author: <a href="https://github.com/lgatto">Laurent Gatto</a> and Robert Stojnić</li>
<li>Original repository: <a href="https://github.com/lgatto/sequences">https://github.com/lgatto/sequences</a>
</li>
<li>More details: <a href="https://github.com/lgatto/sequences/blob/master/DESCRIPTION">https://github.com/lgatto/sequences/blob/master/DESCRIPTION</a>
</li>
<li>Installation from CRAN: <code>install.packages("sequences")</code>
</li>
<li>Installation from github (requires <code>R</code> and <code>C/C++</code> building tools): </li>
</ul>

<div class="highlight highlight-source-c"><pre><span class="pl-en">library</span>(devtools)
install_github(<span class="pl-s"><span class="pl-pds">"</span>sequences<span class="pl-pds">"</span></span>, <span class="pl-s"><span class="pl-pds">"</span>lgatto<span class="pl-pds">"</span></span>)</pre></div>

<h3>
<a id="best-practices-in-bioinformatics-research-open-source-software-and-reproducibility" class="anchor" href="#best-practices-in-bioinformatics-research-open-source-software-and-reproducibility" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Best practices in bioinformatics research: open source software and reproducibility</h3>

<ul>
<li>Description: A short course for a Bioinformatics minor at the
University of Cambridge. What is open science (data, source/code,
access), and how can we enable it? What is reproducible research,
and why do we need it and how can we implement it? The objective is
to familiarise students with concepts and tools of open science and
reproducible research.</li>
<li>Author: <a href="https://github.com/lgatto">Laurent Gatto</a>
</li>
<li>Original repository: <a href="https://github.com/lgatto/open-rr-bioinfo-best-practice">https://github.com/lgatto/open-rr-bioinfo-best-practice</a>
</li>
</ul>

<h3>
<a id="statistics-primer" class="anchor" href="#statistics-primer" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Statistics primer</h3>

<ul>
<li>Description: A short course for a Bioinformatics minor at the
University of Cambridge. Introducing basic concepts in statistics:
experimental design, randomisation, technical and biological
variation, power analysis, hypothesis testing, confidence interval,
what is a p-value, false discovery rate, multiple testing
adjustment, dangers of uninformed statistical practice. The objective
is familiarise students with basic statistical concepts and initiate
them to statistical thinking. They should be able to critically
assess an experimental design and the reporting of a simple
statistical analysis.</li>
<li>Author: <a href="https://github.com/lgatto">Laurent Gatto</a>
</li>
<li>Original repository: <a href="https://github.com/lgatto/statistics-primer">https://github.com/lgatto/statistics-primer</a>
</li>
<li>Slides on
<a href="https://github.com/lgatto/statistics-primer/blob/master/expdes-slides.pdf">experimental desing</a>,
<a href="https://github.com/lgatto/statistics-primer/blob/master/test-slides.pdf">significance testing</a>, 
and
<a href="https://htmlpreview.github.io/?https://github.com/lgatto/statistics-primer/blob/master/03-practical.html">practical</a>.</li>
</ul>
      </section>
      <footer>
        <p>This project is maintained by <a href="https://github.com/lgatto">lgatto</a></p>
        <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a></small></p>
      </footer>
    </div>
    <script src="javascripts/scale.fix.js"></script>
    
  </body>
</html>
