# Material for R/Bioconductor Proteomics Workshop at Stellenbosch University, October 2016

### Material

- [Introduction to R/RStudio](./R-intro/README.md)
- [R/Bioconductor for proteomics](./r4p/README.md)
- [Data visualiation](./vis/README.md)
- [`mzR` and `MSnbase` packages for raw data](./raw/README.md)
- [Indentification data in `mzR`, `mzID`, `MSnID`](./ident/README.md)
- [The `MSnbase` package: `MSnSet`s for quantiative proteomics](./quant/README.md)
- Experimental designs: biological and technical variability, blocking
  and randomization
- Handling missing values in proteomics data
- Differential expression of proteomics data (t-test for continuous
  data, count data, `limma`)
- Multiple testing

### Schedule

|time          |Tue.25.Oct                                                  |Wed.26.Oct                                              |Thu.27.Oct                                                                                    |
|:-------------|:-----------------------------------------------------------|:-------------------------------------------------------|:---------------------------------------------------------------------------------------------|
|9:00 - 10:00  |R/RStudio intro                                             |Identification 1: DB Search and other methods           |Advanced quant: Technology selection, Power, Biofluids                                        |
|10:00 - 11:00 |R                                                           |Practical                                               |Bridget Calder, UCT, demos Skyline                                                            |
|11:00 - 11:30 |Coffee                                                      |Coffee                                                  |Coffee                                                                                        |
|11:30 - 12:30 |R/Bioconductor for proteomics                               |Identification 3: protein inference and controlling FDR |Missing values, bio and tech variability, blocking / randomization (lecture + demo/practical) |
|12:30 - 13:30 |Lunch                                                       |Lunch                                                   |Lunch                                                                                         |
|13:30 - 14:30 |MS data: pwiz and quameter                                  |R/Bioconductor: mzR, mzID, MSnID                        |Differential expression (t-test for continuous data, count data, limma)                       |
|14:30 - 15:30 |Data visualisation                                          |Intro: quantitative proteomics                          |Multiple testing                                                                              |
|15:30 - 16:00 |Coffee                                                      |Coffee                                                  |Coffee                                                                                        |
|16:00 - 17:00 |Practical with R: the mzR and MSnbase packages for raw data |The MSnbase package: MSnSets for quantiative proteomics |Wrap-up                                                                                       |


### Related material and references

* [Teaching material](http://lgatto.github.io/TeachingMaterial/)


* Gatto L. and Christoforou A. *Using R and Bioconductor for
  proteomics data analysis*, Biochim Biophys Acta - Proteins and
  Proteomics, 2013. [PMID:23692960](https://www.ncbi.nlm.nih.gov/pubmed/23692960)
  ([preprint](https://arxiv.org/abs/1305.6559))

* Gatto L, Breckels LM, Naake T, Gibb S. *Visualisation of proteomics
  data using R and Bioconductor*. Proteomics. 2015 Feb 18. doi:
  10.1002/pmic.201400392. [PMID:25690415](http://www.ncbi.nlm.nih.gov/pubmed/25690415).

* Gatto L. and Lilley K.S. MSnbase - an R/Bioconductor package for
  isobaric tagged mass spectrometry data visualisation, processing and
  quantitation, Bioinformatics, 28(2), 288-289, 2012
  [PMID:22113085](https://www.ncbi.nlm.nih.gov/pubmed/22113085)

### License

The content of this repository is made available under the 
[Creative Commons Attribution license](https://creativecommons.org/licenses/by/4.0/). 
The following is a human-readable summary of (and not a substitute for) the 
[full legal text of the CC BY 4.0](https://creativecommons.org/licenses/by/4.0/legalcode).

You are free:

* to **Share** - copy and redistribute the material in any medium or format
* to **Adapt** - remix, transform, and build upon the material

for any purpose, even commercially. The licensor cannot revoke these
freedoms as long as you follow these license terms.

Under the following terms:

**Attribution** - You must give appropriate credit, provide a link to
the license, and indicate if changes were made. You may do so in any
reasonable manner, but not in any way that suggests the licensor
endorses you or your use.
	
No additional restrictions - You may not apply legal terms or
technological measures that legally restrict others from doing
anything the license permits.


	
