# Identifying differentially expressed proteins

```{r, echo = FALSE, warning = FALSE}
set.seed(1)
library("BiocStyle")
suppressPackageStartupMessages(library("pRolocdata"))
suppressPackageStartupMessages(library("msmsTests"))
```

## Running a `t-test` in R

Using the `t.test` function:

```
t.test(x, y = NULL,
       alternative = c("two.sided", "less", "greater"),
       mu = 0, paired = FALSE, var.equal = FALSE,
       conf.level = 0.95, ...)
```

We will focus on **two sample** unpaired t-test, assuming unequal
variances, as this is the most common scenario in proteomics. Using a
**paired test** when appropriate is essential, as it will
substantially increase your test power.

We are going to use the `rnorm` function in this an the next section
to quickly generate normally distributed data. Its inputs are 

- `n`: the number of data points to be generated;
- `mean`: the mean of the normal distribution to draw the data from
  (default is 0);
- `sd`: the standard deviation of the normal distribution to draw the
  data from (default is 1).

### Exercise 

* Generate 200 numbers drawn from a normal distribution of mean 0 and
  standard deviation 1. Verify that the parameters of the randomly
  data are correct. What figure would you use to visualise such data?
  
* Same as above for a normal distribution of mean 2 and standard
  deviation 0.5.

* Compare your values with your neighbour's. Are they identical?

Let's now apply a t-test on two sets of values drawn from identical
and different distributions:

```{r}
t1 <- t.test(rnorm(5), rnorm(5))
t1
```

```{r}
t2 <- t.test(rnorm(5), rnorm(5, mean = 4))
t2
```

What we see above is a pretty output that is convenient to visualise
interactively. The output of the `t.test` is an object of class 
`r class(t2)`, which contains the following values: 

```{r, echo=FALSE}
paste(names(t2), collapse = ", ")
```



We can extract any of these with the `$` accessor

```{r}
t2$p.value
```

## One-sample test 

When using ratio data (as in SILAC or 15N), one would use a **one
sample** t-test.

```{r}
logsilac <- rnorm(3)
t.test(logsilac, mu = 0)
```

### Exercise

Let's use the `mulvey2015` dataset, introduced previously, and focus
on time points 1 and 6. 

1. How can we conveniently use the sample metadata to create the
   relevant subset?


```{r, echo = FALSE, eval = TRUE}
library("pRolocdata")
data(mulvey2015)
time16 <- mulvey2015[, mulvey2015$times %in% c(1, 6)]
## head(exprs(time16))
```

2. Use the `t.test` function to test P48432 for differences in
   time points 1 and 6.

```{r, echo = FALSE, eval = FALSE}
t.test(exprs(time16)[1, time16$time == 1],
       exprs(time16)[1, time16$time != 1])
```

In high throughput biology, we have to repeat our tests over every
feature (transcript, protein, ...). As we are using a programming
language, this is something easy to do.

Below, we use the apply function, that will iterate a function over
all elements of its input.

```{r}
time1 <- time16$time == 1
time6 <- time16$time != 1
## first attempt
pv <- apply(exprs(time16), 1,
            function(x) t.test(x[time1], x[time6]))

## second attempt
pv <- apply(exprs(time16), 1,
            function(x) t.test(x[time1], x[time6])$p.value)
```

We now have calculated a p-value for each of the `r nrow(time16)`
proteins in the data; let's add them to the feature metadata slot.

```{r}
fData(time16)$p.value <-
                apply(exprs(time16), 1,
                      function(x) t.test(x[time1], x[time6])$p.value)
```

```{r}
fData(time16)$fc <-
                apply(exprs(time16), 1,
                      function(x) mean(x[time1])/mean(x[time6]))
fData(time16)$lfc <- log2(fData(time16)$fc)
```

We could also use the `rowttests` function from the `genefilter`
package to run a t-test on all rows at once.

## Visualising results

There are 3 important factors to consider when assessing the results
of a test for differential expression:

* The significance of the test, i.e. the p-values
* The magnitude of the change, i.e. the fold-change
* The (average) intensity of the measurements

```{r}
MAplot(time16)
```

```{r}
plot(fData(time16)$lfc, -log10(fData(time16)$p.value))
```

## Multiple testing

See [this section](../multtest/README.md) for details.

Applying this to our data, we obtain

```{r}
hist(fData(time16)$p.value)
```

```{r}
library("qvalue")
fData(time16)$q.value <- qvalue(fData(time16)$p.value)$qvalue
```

```{r}
plot(fData(time16)$lfc, -log10(fData(time16)$q.value))
summary(fData(time16)$q.value)
```

## Moderated t-tests: `limma`

Two values are used when computing a t statistics: the effect size
(i.e. difference in means) and the sample standard deviations. Any
uncertainty on these values will have a negative impact on the test
results. 

For example, in case of low quantitation values, the difference in
means can fluctuate quite substantially due to random variations and
favour misleading large effect sizes and false positive.

Another example is small sample size, which can bias the estimation of
the sample standard deviations. To address this, **moderated t
statistics** compute variance estimates that are a compromise between
global variance estimate and individual gene-wise variance estimates.

From the `r Biocpkg("limma")` reference:

> This procedure is implemented in the limma software package (Ritchie
> et al., 2015) and the resulting EB [Empirical Bayes] tests have been
> shown to offer improved statistical power and false discovery rate
> (FDR) control relative to the ordinary gene-wise t-tests, especially
> when the sample sizes are small (Kooperberg et al., 2005; Murie et
> al., 2009; Ji and Liu, 2010; Jeanmougin et al., 2010). The limma
> software has been used successfully in thousands of published
> biological studies using data from a variety of genomic
> technologies, especially studies using expression microarrays and
> RNA-seq.

The `r Biocpkg("limma")` comes with extensive documentation, available
with the `limmaUsersGuide()` function. While is doesn't explicitly
mention proteomics, its methodology is applicable to proteomics data.

## Count data

As discussed in the lecture, count data cannot be handled using a test
for continuous data. One could log-transform the data (adding one to
the data to keep 0 counts). Alternatively, using a dedicated count
distribution has proved successful. Methods originally developed for
high throughput sequencing data, have benefited from tremendous
development within the Bioconductor project, and can be readily
applied to proteomics count data.

The `r Biocpkg("msmsTests")` package applies various such count-based
tests on `MSnSet` objects containing spectral counting data. The
package provides a test data `msms.spk`, described as follows in the
manual page:

     A MSnSet with a spectral counts in the expression matrix and a
     treatment factor in the phenoData slot.
     The spectral counts matrix has samples in the columns, and
     proteins in the rows. Each sample consists in 500ng of standard
     yeast lisate spiked with 100, 200, 400 and 600fm of a mix of 48
     equimolar human proteins (UPS1, Sigma-Aldrich). The dataset
     contains a different number of technical replicates of each
     sample.

```{r}
library("msmsTests")
data(msms.spk)
```
### Exercise

* Familiarise yourself with the experimental design of this
  dataset. Hint: look at the phenoData slot.

```{r, echo = FALSE, eval = FALSE}
pData(msms.spk)
table(msms.spk$treat)
```

* How many samples and proteins are there in the data

```{r, eval = FALSE, echo=FALSE}
dim(msms.spk)
```

* Look at the distribution of all proteins and compare it to the spike
  in proteins. The spikes all contain the suffix `"HUMAN"` that can be
  extracted with the grep function.
  
```{r, eval = FALSE, echo = FALSE}
boxplot(exprs(msms.spk))  
spks <- grep("HUMAN", featureNames(msms.spk))  
boxplot(exprs(msms.spk[spks, ]))
```  

We are going to model the data according to the negative-binomial
distribution, using the implementation of the `r Biocpkg("edgeR")`
package, which uses an Empirical Bayes method to share information
across features and is this particularly relevant with a restricted
number of replicates. We will focus on the `U200` and `U600`
conditions.

```{r}
e <- msms.spk[, msms.spk$treat %in% c("U200", "U600")]
table(e$treat)
```

We now also need to remove proteins that are left with only 0.

```{r}
e <- filterZero(e, pNA = 0.99)
```

We can run the NB spectral counts differential test using the
`msms.edgeR`, providing

* an `MSnSet`, here `e`
* an alternative and null hypothesis, `H1` (there is a *treatment*
  effect) and `H0` (there is no effect, the expression is essentially
  constant)
* the groups, `e$treat`
* a column-wise scaling offset (optional)


```{r}
H0 <- "y ~ 1"
H1 <- "y ~ treat"

## normalising condition
div <- colSums(exprs(e))

res <- msms.edgeR(e, H1, H0,
                  fnm = "treat",
                  div = div)
head(res)
```

### Exercise

* Inspect the p-values distribution and, if relevant, adjust as
  demonstrated above.

```{r, echo = FALSE, eval = FALSE}
hist(res$p.value, breaks = 50)
library("multtest")
adj <- mt.rawp2adjp(res$p.value)
res$BH <- adj$adjp[order(adj$index), "BH"]
## with(res, plot(LogFC, -log10(BH)))
```

* Visualise the results on a volcano plot

```{r, eval = FALSE, echo = FALSE}
sig <- res$BH < 0.01
plot(res$LogFC, -log10(res$BH),
     col = ifelse(sig, "red", "black"),
     pch = ifelse(grepl("HUMAN", featureNames(e)),
                  19, 1))
```

* Estimate the number true/false positives and negatives and an alpha
  level of 0.01.


```{r, echo = FALSE}
sig <- res$BH < 0.01
TP <- length(grep("HUMAN", rownames(res)[sig]))
FP <- sum(sig) - TP
FN <- length(grep("HUMAN", rownames(res)[!sig]))
TN <- sum(!sig) - FN
stopifnot(TN + FP + FN + TP == nrow(e))
tab <- data.frame(TP, FP, TN, FN)
```

## Other packages

* `r Biocpkg("MSstats")` for various statistical analyses
* Isobaric tagging (iTRAQ and TMT): `r Biocpkg("isobar")`
* Label-free: `r CRANpkg("aLFQ")` and `r CRANpkg("protiq")`

